tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/GraphQL/HotChocolate/HotChocolateCommon.cs

```diff
@@ -0,0 +1,163 @@
+using Datadog.Trace.Configuration;
+using Datadog.Trace.DuckTyping;
+using Datadog.Trace.Logging;
+
+namespace Datadog.Trace.ClrProfiler.AutoInstrumentation.GraphQL.HotChocolate
+{
+    internal class HotChocolateCommon : GraphQLCommonBase
+    {
+        internal const string IntegrationName = nameof(Configuration.IntegrationId.HotChocolate);
+        internal const IntegrationId IntegrationId = Configuration.IntegrationId.HotChocolate;
+
+        private const string ServiceName = "hotchocolate";
+        internal const string ErrorType = "HotChocolate.Error";
+
+        private static readonly IDatadogLogger Log = DatadogLogging.GetLoggerFor(typeof(HotChocolateCommon));
+
+        internal static Scope CreateScopeFromExecuteAsync(Tracer tracer, IQueryRequest request)
+        {
+            if (!tracer.Settings.IsIntegrationEnabled(IntegrationId))
+            {
+                // integration disabled, don't create a scope, skip this trace
+                return null;
+            }
+
+            Scope scope = null;
+            try
+            {
+                var queryOperationName = request.OperationName;
+                var source = request.Query?.ToString();
+                var operationType = "Uncompleted";
+                scope = CreateScopeFromExecuteAsync(tracer, IntegrationId, new HotChocolateTags(), ServiceName, queryOperationName, source, operationType);
+            }
+            catch (Exception ex)
+            {
+                Log.Error(ex, "Error creating or populating scope.");
+            }
+
+            return scope;
+        }
+
+        internal static Scope UpdateScopeFromExecuteAsync(Tracer tracer, IOperationContext context)
+        {
+            if (!tracer.Settings.IsIntegrationEnabled(IntegrationId))
+            {
+                // integration disabled, don't create a scope, skip this trace
+                return null;
+            }
+
+            var scope = tracer.InternalActiveScope;
+            var span = scope?.Span;
+            if (span == null || span.OperationName != ExecuteOperationName)
+            {
+                // not in a Hotchocolate execution span
+                return null;
+            }
+
+            try
+            {
+                var operation = context.Operation;
+                var operationType = operation.OperationType.ToString();
+                var operationName = span.GetTag(Trace.Tags.GraphQLOperationName);
+                span.ResourceName = $"{operationType} {operationName ?? "operation"}";
+                span.SetTag(Trace.Tags.GraphQLOperationType, operationType);
+            }
+            catch (Exception ex)
+            {
+                Log.Error(ex, "Error creating or populating scope.");
+            }
+
+            return scope;
+        }
+
+        internal static void RecordExecutionErrorsIfPresent(Span span, string errorType, System.Collections.IEnumerable errors)
+        {
+            var executionErrors = GetList(errors);
```

Comment:
Hmmm, I think we should avoid calling GetList here if possible 🤔 I think we could do the DuckTyping inside `ConstructErrorMessage`, that would avoid the extra list creation?